pkg/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,253 @@
+		ev.stats.waitEnqueue.End(true)
+		ev.stats.waitConsumeOffQueue.Start()
+		return ev.eventResults
+	}
+}
+
+func (ev *Event) printStats() {
+	log.WithFields(logrus.Fields{
+		"eventType":                    reflect.TypeOf(ev.Metadata).String(),
+		"eventHandlingDuration":        ev.stats.durationStat.Total(),
+		"eventEnqueueWaitTime":         ev.stats.waitEnqueue.Total(),
+		"eventConsumeOffQueueWaitTime": ev.stats.waitConsumeOffQueue.Total(),
+	}).Debug("EventQueue event processing statistics")
+}
+
+// Run consumes events that have been queued for this EventQueue. It
+// is presumed that the eventQueue is a buffered channel with a length of one
+// (i.e., only one event can be processed at a time). All business logic for
+// handling queued events is contained within this function. If the Metadata
+// within an Event does not implement the EventHandler interface, the function
+// will log an error, and the Event will not be processed. If the event queue is
+// closed, then all events which were queued up, but not processed, are
+// cancelled. Any event which is currently being processed will not be
+// cancelled. It is assumed that the caller handles both cases for an Event
+// (cancel, or result of event) gracefully.
+func (q *EventQueue) Run() {
+	go q.eventQueueOnce.Do(func() {
+		for {
+			select {
+			// Receive next event. No other goroutine or process should consume
+			// events off of this channel!
+			case ev := <-q.events:
+				{
+					ev.stats.waitConsumeOffQueue.End(true)
+					eventHandler := ev.Metadata.(EventHandler)
+					ev.stats.durationStat.Start()
+					eventHandler.Handle(ev.eventResults)
+					// Always indicate success for now.s
+					ev.stats.durationStat.End(true)
+					// Ensures that no more results can be sent as the event has
+					// already been processed.
+					ev.printStats()
+					close(ev.eventResults)
+				}
+			// Cancel all events that were not yet consumed.
+			case <-q.close:
+				{
+					// Drain queue of all events.
+					for {
+						select {
+						case drainEvent := <-q.events:
+							drainEvent.stats.waitConsumeOffQueue.End(false)
+							close(drainEvent.cancelled)
+							close(drainEvent.eventResults)
+							drainEvent.printStats()
+						default:
+							// No more events are in events channel, so we can close
+							// it and exit. It is guaranteed that no more events
+							// will be queued onto the events channel because
+							// the close channel has been closed. See Enqueue.
+							close(q.drained)
+							close(q.events)
+							return
+						}
+					}
+				}
+			}
+		}
+	})
+}
+
+// Stop stops any further events from being processed by the EventQueue. Any
+// event which is currently being processed by the EventQueue will continue to
+// run. All other events waiting to be processed, and all events that may be
+// enqueued will not be processed by the event queue; they will be cancelled.
+// If the queue has already been stopped, this is a no-op.
+func (q *EventQueue) Stop() {
```

Comment:
Multiple calls to Stop() can cause panic on channel closed for `q.close`

Reply:
Are you saying the panic occurring for the following case?

call to `q.Stop`() and immediately after `q.Stop()` before `q.close` before it is closed in the `default` statement for the first invocation of `q.Stop()`?

Reply:
I'll wrap this in a `sync.Once`